File: 0046-permutations/0046-permutations.java
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    private int[] nums;

    private void permutation(List<Integer> path, int idx){
        if(idx==nums.length){
            res.add(path);
            return;
        }

        for(int i=idx; i<nums.length; i++){
            Collections.swap(path, idx, i);
            permutation(new ArrayList<>(path), idx+1);
            Collections.swap(path, i, idx);
        }
    }

    public List<List<Integer>> permute(int[] nums) {
        this.nums = nums;

        List<Integer> path = new ArrayList<>();
        for(int i:nums){
            path.add(i);
        }
        permutation(path, 0);
        return res;
    }
}

File: 0046-permutations/0046-permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = len(nums)
        path = nums
        res = []

        def permute(path, idx):
            if idx == n:
                res.append(list(path))
                return

            for i in range(idx, n):
                path[idx], path[i] = path[i], path[idx]
                permute(path, idx+1)
                path[idx], path[i] = path[i], path[idx]

        
        permute(path, 0)
        return res

